brave/browser/net/proxy/proxy_config_service_tor.cc

```diff
@@ -4,32 +4,136 @@
 
 #include "brave/browser/net/proxy/proxy_config_service_tor.h"
 
+#include <memory>
+#include <utility>
 #include <vector>
 
+#include "base/path_service.h"
 #include "base/strings/string_number_conversions.h"
 #include "base/strings/string_util.h"
+#include "chrome/common/chrome_paths.h"
 #include "crypto/random.h"
-#include "url/gurl.h"
+#include "net/proxy/proxy_service.h"
+#include "net/url_request/url_request_context.h"
+#include "services/service_manager/public/cpp/connector.h"
 
 namespace net {
 
+using content::BrowserThread;
 const int kTorPasswordLength = 16;
 
-ProxyConfigServiceTor::ProxyConfigServiceTor(const GURL& url)
-  : url_(url) {
-    config_.proxy_rules().ParseFromString(url_.spec());
+ProxyConfigServiceTor::ProxyConfigServiceTor(const std::string tor_path,
+  const std::string tor_proxy) {
+    if (tor_path.length() && tor_proxy.length()) {
+      tor_path_ = tor_path;
+      url::Parsed url;
+      url::ParseStandardURL(
+        tor_proxy.c_str(),
+        std::min(tor_proxy.size(),
+                 static_cast<size_t>(std::numeric_limits<int>::max())),
+        &url);
+      if (url.scheme.is_valid()) {
+        scheme_ =
+          std::string(tor_proxy.begin() + url.scheme.begin,
+                      tor_proxy.begin() + url.scheme.begin + url.scheme.len);
+      }
+      if (url.host.is_valid()) {
+        host_ =
+          std::string(tor_proxy.begin() + url.host.begin,
+                      tor_proxy.begin() + url.host.begin + url.host.len);
+      }
+      if (url.port.is_valid()) {
+        port_ =
+          std::string(tor_proxy.begin() + url.port.begin,
+                      tor_proxy.begin() + url.port.begin + url.port.len);
+      }
+
+      BrowserThread::PostTask(
+          BrowserThread::PROCESS_LAUNCHER, FROM_HERE,
+          base::Bind(&ProxyConfigServiceTor::LaunchTorProcess,
+                     base::Unretained(this)));
+    }
+    config_.proxy_rules().ParseFromString(std::string(scheme_ + "://" + host_
+      + ":" + port_));
+}
+
+void ProxyConfigServiceTor::OnTorCrashed(int32_t pid) {
+  LOG(ERROR) << "Tor Process(" << pid << ") Crashed";
+  BrowserThread::PostTask(
+      BrowserThread::PROCESS_LAUNCHER, FROM_HERE,
+      base::Bind(&ProxyConfigServiceTor::LaunchTorProcess,
+                 base::Unretained(this)));
 }
 
 ProxyConfigServiceTor::~ProxyConfigServiceTor() {}
 
+void ProxyConfigServiceTor::LaunchTorProcess() {
+  content::ServiceManagerConnection::GetForProcess()
```

Comment:
put `GetConnector` and `set_connection_error_handler` before `LaunchTorProcess` inside constructor.
This will cause duplicate services running after tor process killed